### PR TITLE
Ajax API: Get Coupon by ID

### PR DESCRIPTION
**Pull request by mi-prakash**
Wp Ajax action
**Payload**
- `action` - tutor_coupon_details
- `coupon_id` - {ID}

image

Task URL
https://app.clickup.com/t/86eprdrru

### CODE REVIEW

File: classes/Tutor.php [hunk 1]
```diff
@@ -928,18 +928,23 @@ public static function create_database() {
...
 			id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT,
 			coupon_status VARCHAR(50),
+			coupon_type VARCHAR(100) DEFAULT 'code', -- coupon type 'code' or 'automatic'
 			coupon_code VARCHAR(50) NOT NULL,
 			coupon_title VARCHAR(255) NOT NULL,
 			coupon_description TEXT,
 			discount_type ENUM('percentage', 'flat') NOT NULL,
 			discount_amount DECIMAL(13, 2) NOT NULL,
-			total_usage_limit INT(10) UNSIGNED,
-			per_user_usage_limit INT(10) UNSIGNED,
```
Comment (shewa12) [inline classes/Tutor.php:937]: Why removed?

File: classes/Tutor.php [hunk 2]
```diff
@@ -928,18 +928,23 @@ public static function create_database() {
...
-			coupon_type VARCHAR(100) NOT NULL, -- product specific or category or all
+			applies_to VARCHAR(100) DEFAULT 'all_courses_and_bundles', -- possible values 'all_courses_and_bundles', 'all_courses', 'all_bundles', 'specific_courses', 'specific_bundles', 'specific_category'
+			total_usage_limit INT(10) UNSIGNED DEFAULT 0,
+			is_limit_to_one_use_per_customer TINYINT(4) UNSIGNED DEFAULT 0,
```
Comment (shewa12) [inline classes/Tutor.php:939]: Why this column `is_limit_to_one_use_per_customer `?

File: classes/Tutor.php [hunk 3]
```diff
@@ -928,18 +928,23 @@ public static function create_database() {
...
+			purchase_requirement VARCHAR(50) DEFAULT 'no_minimum', -- possible values 'no_minimum', 'minimum_purchase', 'minimum_quantity'
```
Comment (shewa12) [inline classes/Tutor.php:940]: I think this should be
`coupon_requirement VARCHAR(50) DEFAULT 'na', -- possible values 'na', 'min_purchase', 'min_quantity'`
And coupon_requirement_value

File: classes/Utils.php [hunk 4]
```diff
@@ -902,7 +902,7 @@ public function get_course_completed_percent( $course_id = 0, $user_id = 0, $get
 					 */
 					$is_completed = apply_filters( 'tutor_is_zoom_lesson_done', false, $content->ID, $user_id );
 					if ( $is_completed ) {
-						$completed_count++;
```
Comment (shewa12) [inline classes/Utils.php:905]: Why this change? Please make sure it will not affect any other areas!

File: classes/Utils.php [hunk 5]
```diff
@@ -912,7 +912,7 @@ public function get_course_completed_percent( $course_id = 0, $user_id = 0, $get
 					 */
 					$is_completed = apply_filters( 'tutor_google_meet_lesson_done', false, $content->ID, $user_id );
 					if ( $is_completed ) {
-						$completed_count++;
```
Comment (shewa12) [inline classes/Utils.php:915]: Same here

File: classes/Utils.php [hunk 6]
```diff
@@ -7345,7 +7343,7 @@ public function get_course_prev_next_contents_by_id( $content_id = 0 ) {
 						$next_id = $ids[ $next_i ];
 					}
 				}
-				$i++;
```
Comment (shewa12) [inline classes/Utils.php:7348]: same here?

File: classes/Utils.php [hunk 7]
```diff
@@ -9045,26 +9043,26 @@ public function count_completed_contents_by_topic( int $topic_id ): array {
 					case $lesson_post_type:
 						$is_lesson_completed = $this->is_completed_lesson( $content->ID, $user_id );
 						if ( $is_lesson_completed ) {
-							$completed++;
```
Comment (shewa12) [inline classes/Utils.php:9048]: Please check every post & pre increment

**Review by shewa12 — CHANGES_REQUESTED** [review]
Please check the reviews

File: classes/Tutor.php (hunk 1 above)
Comment (mi-prakash) [reply classes/Tutor.php:937]: these are based on the design
there is not user limit in the design
only a check mark to make the user usage limit by 1

File: classes/Tutor.php (hunk 2 above)
Comment (mi-prakash) [reply classes/Tutor.php:939]: described in previous comment

File: classes/Utils.php (hunk 4 above)
Comment (mi-prakash) [reply classes/Utils.php:905]: I need to check these. My ide formatter might formatted this. I'll check and fix this

File: classes/Utils.php (hunk 5 above)
Comment (mi-prakash) [reply classes/Utils.php:915]: same here

File: classes/Utils.php (hunk 6 above)
Comment (mi-prakash) [reply classes/Utils.php:7348]: same here

File: classes/Utils.php (hunk 7 above)
Comment (mi-prakash) [reply classes/Utils.php:9048]: same here

File: classes/Tutor.php (hunk 2 above)
Comment (mi-prakash) [reply classes/Tutor.php:939]: i'll change the name from  `is_limit_to_one_use_per_customer` -> `is_one_use_per_user`

File: classes/Tutor.php (hunk 3 above)
Comment (mi-prakash) [reply classes/Tutor.php:940]: i believe `no_minimum` makes more suitable event for the api response for the team to work the data with

File: classes/Utils.php [hunk 8]
```diff
@@ -1119,11 +1119,12 @@ public function is_course_purchasable( $course_id = 0 ) {
...
 	 * @since 1.0.0
 	 *
-	 * @param int $course_id course ID.
+	 * @param int  $course_id course ID.
+	 * @param bool $apply_filter false if you don't want to apply the filter on return.
 	 *
 	 * @return null|string
 	 */
-	public function get_course_price( $course_id = 0 ) {
+	public function get_course_price( $course_id = 0, $apply_filter = true ) {
```
Comment (shewa12) [inline classes/Utils.php:1127]: Need to get this method back to the previous state

**Review by shewa12 — APPROVED** [review]
